substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -941,39 +928,36 @@ impl<T: Config> Pallet<T> {
         ContractBillingInformationByID::<T>::insert(report.contract_id, &contract_billing_info);
     }
 
-    pub fn _bill_contract_for_block(
-        contract_id: u64,
-        block: T::BlockNumber,
-    ) -> DispatchResultWithPostInfo {
+    pub fn _bill_contract(contract_id: u64) -> DispatchResultWithPostInfo {
         if !Contracts::<T>::contains_key(contract_id) {
+            log::debug!("cleaning up deleted contract from storage");
+
+            let now = <frame_system::Pallet<T>>::block_number().saturated_into::<u64>();
+            let index = now % BillingFrequency::<T>::get();
+
+            // Remove contract from billing list
+            let mut contracts = ContractsToBillAt::<T>::get(index);
+            contracts.retain(|&c| c != contract_id);
+            ContractsToBillAt::<T>::insert(index, contracts);
+
             return Ok(().into());
         }
         let mut contract = Contracts::<T>::get(contract_id).ok_or(Error::<T>::ContractNotExists)?;
 
-        log::info!(
-            "billing contract with id {:?} at block {:?}",
-            contract_id,
-            block
-        );
+        log::info!("billing contract with id {:?}", contract_id);
+
         // Try to bill contract
         match Self::bill_contract(&mut contract) {
             Ok(_) => {
-                log::info!(
-                    "successfully billed contract with id {:?} at block {:?}",
-                    contract_id,
-                    block
-                );
+                log::info!("successfully billed contract with id {:?}", contract_id,);
```

Comment:
It's not a big deal but can we maybe move this to the bill_contract function? Because if the grace period is started then this function will return Ok while the billing did not take place. And this message then suggests that the contract was successfully billed.

Reply:
Better solution would be is to actually return an error if the contract is in grace period no?

Reply:
That's also possible yes